# Tasca M3 T02

## Exercici 1

- Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un Pandas Dataframe. 
- Explora les dades que conté, explica breument quines variables hi ha i queda’t únicament amb les columnes que consideris rellevants. Justifica la teva elecció.

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

retraso_vuelos = pd.read_csv('DelayedFlights.csv', sep= ',', index_col= 0)

retraso_vuelos

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0        2008      1           3          4   2003.0        1955   2211.0   
1        2008      1           3          4    754.0         735   1002.0   
2        2008      1           3          4    628.0         620    804.0   
4        2008      1           3          4   1829.0        1755   1959.0   
5        2008      1           3          4   1940.0        1915   2121.0   
...       ...    ...         ...        ...      ...         ...      ...   
7009710  2008     12          13          6   1250.0        1220   1617.0   
7009717  2008     12          13          6    657.0         600    904.0   
7009718  2008     12          13          6   1007.0         847   1149.0   
7009726  2008     12          13          6   1251.0        1240   1446.0   
7009727  2008     12          13          6   1110.0        1103   1413.0   

         CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  Cancelled  \
0              2225            WN        335  ...    4.0      8.0          0   
1              1000            WN       3231  ...    5.0     10.0          0   
2               750            WN        448  ...    3.0     17.0          0   
4              1925            WN       3920  ...    3.0     10.0          0   
5              2110            WN        378  ...    4.0     10.0          0   
...             ...           ...        ...  ...    ...      ...        ...   
7009710        1552            DL       1621  ...    9.0     18.0          0   
7009717         749            DL       1631  ...   15.0     34.0          0   
7009718        1010            DL       1631  ...    8.0     32.0          0   
7009726        1437            DL       1639  ...   13.0     13.0          0   
7009727        1418            DL       1641  ...    8.0     11.0          0   

         CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
0                       N         0           NaN          NaN      NaN   
1                       N         0           NaN          NaN      NaN   
2                       N         0           NaN          NaN      NaN   
4                       N         0           2.0          0.0      0.0   
5                       N         0           NaN          NaN      NaN   
...                   ...       ...           ...          ...      ...   
7009710                 N         0           3.0          0.0      0.0   
7009717                 N         0           0.0         57.0     18.0   
7009718                 N         0           1.0          0.0     19.0   
7009726                 N         0           NaN          NaN      NaN   
7009727                 N         0           NaN          NaN      NaN   

         SecurityDelay  LateAircraftDelay  
0                  NaN                NaN  
1                  NaN                NaN  
2                  NaN                NaN  
4                  0.0               32.0  
5                  NaN                NaN  
...                ...                ...  
7009710            0.0               22.0  
7009717            0.0                0.0  
7009718            0.0               79.0  
7009726            NaN                NaN  
7009727            NaN                NaN  

[1936758 rows x 29 columns]

### Retraso de vuelo EDA (Análisis exploratorio de datos)

Acerca del conjunto de datos Contexto La Oficina de Estadísticas de Transporte del Departamento de Transporte de EE. UU. (DOT) realiza un seguimiento de la puntualidad de los vuelos nacionales operados por grandes compañías aéreas. La información resumida sobre la cantidad de vuelos a tiempo, retrasados, cancelados y desviados se publica en el Informe mensual de consumidores de viajes aéreos del DOT y en este conjunto de datos de retrasos y cancelaciones de vuelos de 2008.

Se realizará un análisis del conjunto de datos proporcionado, este contiene hasta 1.936.758 vuelos internos diferentes en los EE. UU. para 2008 y sus causas de retraso, desvío y cancelación. Los datos provienen de la Oficina de Estadísticas de Transporte (BTS) del Departamento de Transporte de EE. UU. (DOT). Los resultados son tentativos.

Este conjunto de datos está compuesto por las siguientes variables: 
1.	Year: 2008.
2.	Month: 1-12.
3.	DayofMonth: 1-31.  
4.	DayOfWeek: 1 (lunes) - 7 (domingo). 
5.	DepTime: hora de salida real (local, hhmm).
6.	CRSDepTime: hora de salida programada (local, hhmm). 
7.	ArrTime: hora de llegada real (local, hhmm). 
8.	CRSArrTime: hora de llegada programada (local, hhmm). 
9.	UniqueCarrier: código único de la compañía aérea.
10.	FlightNum: número de vuelo. 
11.	TailNum: número de cola del avión: matrícula de la aeronave, identificador único de la aeronave.
12.	ActualElapsedTime: en minutos.
13.	CRSElapsedTime: en minutos.
14.	AirTime: en minutos.  
15.	ArrDelay: demora en la llegada, en minutos: Un vuelo se cuenta como "a tiempo" si se realizó con menos de 15 minutos de retraso con respecto a la hora programada que se muestra en los Sistemas de Reservas Computarizados (CRS) de las aerolíneas. 
16.	DepDelay: retraso en la salida, en minutos. 
17.	Origin: código de origen del aeropuerto IATA. 
18.	Dest: código de destino de aeropuerto IATA.
19.	Distance: en millas.
20.	TaxiIn: taxi a tiempo en minutos.
21.	TaxiOut: taxi fuera de tiempo en minutos. 
22.	Cancelled: *era el vuelo cancelado.
23.	CancellationCode: motivo de la cancelación (A = transportista, B = clima , C = NAS, D = seguridad). 
24.	Diverted:1 = sí, 0 = no. 
25.	CarrierDelay: en minutos: El retraso del transportista está bajo el control del transportista aéreo. Ejemplos de sucesos que pueden determinar el retraso del transportista son: limpieza de la aeronave, daños en la aeronave, espera de la llegada de pasajeros o tripulantes de conexión, equipaje, colisión con aves, carga de carga, catering, computadora, interrupción del equipo del transportista, legalidad de la tripulación (descanso del piloto o asistente) , daños por mercancías peligrosas, inspección de ingeniería, abastecimiento de combustible, manejo de pasajeros discapacitados, tripulación retrasada, servicio de lavabos, mantenimiento, sobreventas, servicio de agua potable, retiro de pasajeros ingobernables, embarque o asiento lentos, almacenamiento de equipaje de mano, retrasos por peso y equilibrio. 
26.	WeatherDelay: en minutos: el retraso del tiempo es causado por condiciones climáticas extremas o peligrosas que se pronostican o se manifiestan en el punto de partida, en ruta o en el punto de llegada.
27.	NASDelay: en minutos: el retraso que está bajo el control del Sistema Nacional del Espacio Aéreo (NAS) puede incluir: condiciones climáticas no extremas, operaciones aeroportuarias, volumen de tráfico pesado, control de tráfico aéreo, etc. 
28.	SecurityDelay: en minutos: el retraso de seguridad es causado por la evacuación de una terminal o vestíbulo, volver a abordar una aeronave debido a una violación de la seguridad, equipos de control inoperativos y/o largas filas de más de 29 minutos en las áreas de proyección.
29.	LateAircraftDelay: en minutos: Retraso en la llegada a un aeropuerto debido a la llegada tardía de la misma aeronave a un aeropuerto anterior. El efecto dominó de un retraso anterior en los aeropuertos posteriores se denomina propagación del retraso. 

In [2]:
retraso_vuelos.head()

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0  2008      1           3          4   2003.0        1955   2211.0   
1  2008      1           3          4    754.0         735   1002.0   
2  2008      1           3          4    628.0         620    804.0   
4  2008      1           3          4   1829.0        1755   1959.0   
5  2008      1           3          4   1940.0        1915   2121.0   

   CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  Cancelled  \
0        2225            WN        335  ...    4.0      8.0          0   
1        1000            WN       3231  ...    5.0     10.0          0   
2         750            WN        448  ...    3.0     17.0          0   
4        1925            WN       3920  ...    3.0     10.0          0   
5        2110            WN        378  ...    4.0     10.0          0   

   CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
0                 N         0           NaN          NaN      NaN   
1                 N         0           NaN          NaN      NaN   
2                 N         0           NaN          NaN      NaN   
4                 N         0           2.0          0.0      0.0   
5                 N         0           NaN          NaN      NaN   

   SecurityDelay  LateAircraftDelay  
0            NaN                NaN  
1            NaN                NaN  
2            NaN                NaN  
4            0.0               32.0  
5            NaN                NaN  

[5 rows x 29 columns]

In [3]:
retraso_vuelos.describe()

Year         Month    DayofMonth     DayOfWeek       DepTime  \
count  1936758.0  1.936758e+06  1.936758e+06  1.936758e+06  1.936758e+06   
mean      2008.0  6.111106e+00  1.575347e+01  3.984827e+00  1.518534e+03   
std          0.0  3.482546e+00  8.776272e+00  1.995966e+00  4.504853e+02   
min       2008.0  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%       2008.0  3.000000e+00  8.000000e+00  2.000000e+00  1.203000e+03   
50%       2008.0  6.000000e+00  1.600000e+01  4.000000e+00  1.545000e+03   
75%       2008.0  9.000000e+00  2.300000e+01  6.000000e+00  1.900000e+03   
max       2008.0  1.200000e+01  3.100000e+01  7.000000e+00  2.400000e+03   

         CRSDepTime       ArrTime    CRSArrTime     FlightNum  \
count  1.936758e+06  1.929648e+06  1.936758e+06  1.936758e+06   
mean   1.467473e+03  1.610141e+03  1.634225e+03  2.184263e+03   
std    4.247668e+02  5.481781e+02  4.646347e+02  1.944702e+03   
min    0.000000e+00  1.000000e+00  0.000000e+00  1.000000e+00   
25%    1.135000e+03  1.316000e+03  1.325000e+03  6.100000e+02   
50%    1.510000e+03  1.715000e+03  1.705000e+03  1.543000e+03   
75%    1.815000e+03  2.030000e+03  2.014000e+03  3.422000e+03   
max    2.359000e+03  2.400000e+03  2.400000e+03  9.742000e+03   

       ActualElapsedTime  ...      Distance        TaxiIn       TaxiOut  \
count       1.928371e+06  ...  1.936758e+06  1.929648e+06  1.936303e+06   
mean        1.333059e+02  ...  7.656862e+02  6.812975e+00  1.823220e+01   
std         7.206007e+01  ...  5.744797e+02  5.273595e+00  1.433853e+01   
min         1.400000e+01  ...  1.100000e+01  0.000000e+00  0.000000e+00   
25%         8.000000e+01  ...  3.380000e+02  4.000000e+00  1.000000e+01   
50%         1.160000e+02  ...  6.060000e+02  6.000000e+00  1.400000e+01   
75%         1.650000e+02  ...  9.980000e+02  8.000000e+00  2.100000e+01   
max         1.114000e+03  ...  4.962000e+03  2.400000e+02  4.220000e+02   

          Cancelled      Diverted  CarrierDelay  WeatherDelay      NASDelay  \
count  1.936758e+06  1.936758e+06  1.247488e+06  1.247488e+06  1.247488e+06   
mean   3.268348e-04  4.003598e-03  1.917940e+01  3.703571e+00  1.502164e+01   
std    1.807562e-02  6.314722e-02  4.354621e+01  2.149290e+01  3.383305e+01   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
50%    0.000000e+00  0.000000e+00  2.000000e+00  0.000000e+00  2.000000e+00   
75%    0.000000e+00  0.000000e+00  2.100000e+01  0.000000e+00  1.500000e+01   
max    1.000000e+00  1.000000e+00  2.436000e+03  1.352000e+03  1.357000e+03   

       SecurityDelay  LateAircraftDelay  
count   1.247488e+06       1.247488e+06  
mean    9.013714e-02       2.529647e+01  
std     2.022714e+00       4.205486e+01  
min     0.000000e+00       0.000000e+00  
25%     0.000000e+00       0.000000e+00  
50%     0.000000e+00       8.000000e+00  
75%     0.000000e+00       3.300000e+01  
max     3.920000e+02       1.316000e+03  

[8 rows x 24 columns]

### Tipos de datos

In [4]:
retraso_vuelos.dtypes


Year                   int64
Month                  int64
DayofMonth             int64
DayOfWeek              int64
DepTime              float64
CRSDepTime             int64
ArrTime              float64
CRSArrTime             int64
UniqueCarrier         object
FlightNum              int64
TailNum               object
ActualElapsedTime    float64
CRSElapsedTime       float64
AirTime              float64
ArrDelay             float64
DepDelay             float64
Origin                object
Dest                  object
Distance               int64
TaxiIn               float64
TaxiOut              float64
Cancelled              int64
CancellationCode      object
Diverted               int64
CarrierDelay         float64
WeatherDelay         float64
NASDelay             float64
SecurityDelay        float64
LateAircraftDelay    float64
dtype: object

### Eliminación de variables

**Ignoraré las siguientes columnas:** 
- 1) Año (Year), debido a que el conjunto de datos es para 2008, 
- 2) Número de vuelo (FlightNum),
- 3) Número de cola del avión (TailNum) ya que al ser datos específicos de los aviones, no serán necesarios en este análisis.
- 4) Origen (Origin) y destino (Dest) no serán necesarios porque analizaremos los retrasos. 


In [5]:
retraso_vuelos.drop(['Year', 'FlightNum', 'TailNum', 'Origin', 'Dest'], axis=1, inplace=True)

retraso_vuelos.head()

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
0      1           3          4   2003.0        1955   2211.0        2225   
1      1           3          4    754.0         735   1002.0        1000   
2      1           3          4    628.0         620    804.0         750   
4      1           3          4   1829.0        1755   1959.0        1925   
5      1           3          4   1940.0        1915   2121.0        2110   

  UniqueCarrier  ActualElapsedTime  CRSElapsedTime  ...  TaxiIn  TaxiOut  \
0            WN              128.0           150.0  ...     4.0      8.0   
1            WN              128.0           145.0  ...     5.0     10.0   
2            WN               96.0            90.0  ...     3.0     17.0   
4            WN               90.0            90.0  ...     3.0     10.0   
5            WN              101.0           115.0  ...     4.0     10.0   

   Cancelled  CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
0          0                 N         0           NaN           NaN      NaN   
1          0                 N         0           NaN           NaN      NaN   
2          0                 N         0           NaN           NaN      NaN   
4          0                 N         0           2.0           0.0      0.0   
5          0                 N         0           NaN           NaN      NaN   

   SecurityDelay  LateAircraftDelay  
0            NaN                NaN  
1            NaN                NaN  
2            NaN                NaN  
4            0.0               32.0  
5            NaN                NaN  

[5 rows x 24 columns]

In [39]:
retraso_vuelos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1247483 entries, 4 to 7009718
Data columns (total 29 columns):
 #   Column             Non-Null Count    Dtype  
---  ------             --------------    -----  
 0   Month              1247483 non-null  int64  
 1   DayofMonth         1247483 non-null  int64  
 2   DayOfWeek          1247483 non-null  int64  
 3   DepTime            1247483 non-null  float64
 4   CRSDepTime         1247483 non-null  int64  
 5   ArrTime            1247483 non-null  float64
 6   CRSArrTime         1247483 non-null  int64  
 7   UniqueCarrier      1247483 non-null  object 
 8   ActualElapsedTime  1247483 non-null  float64
 9   CRSElapsedTime     1247483 non-null  float64
 10  AirTime            1247483 non-null  float64
 11  ArrDelay           1247483 non-null  float64
 12  DepDelay           1247483 non-null  float64
 13  Distance           1247483 non-null  int64  
 14  TaxiIn             1247483 non-null  float64
 15  TaxiOut            1247483 non-n

## Exercici 2

Fes un informe complet del dataset:

- Resumeix estadísticament el dataset i les columnes d’interès. 
- Fes una anàlisi estadístic del que consideris rellevant.
- Troba quantes dades faltants hi ha per columna.
- Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...).
- Fes una taula de les aerolínies amb més endarreriments acumulats.
- Quins són els vols més llargs? I els més endarrerits? Busca les rutes més llargues i les que acumulen més retards.
- Aporta allò que consideris rellevant.


### Resumen Estadístico

In [6]:
# Método para eliminar forma exponencial
retraso_vuelos[['ActualElapsedTime', 'CRSElapsedTime', 'AirTime','ArrDelay', 'DepDelay', 'Distance']].describe().apply(lambda s: s.apply('{0:.5f}'.format)) 

ActualElapsedTime CRSElapsedTime        AirTime       ArrDelay  \
count     1928371.00000  1936560.00000  1928371.00000  1928371.00000   
mean          133.30586      134.30274      108.27715       42.19988   
std            72.06007       71.34144       68.64261       56.78472   
min            14.00000      -25.00000        0.00000     -109.00000   
25%            80.00000       82.00000       58.00000        9.00000   
50%           116.00000      116.00000       90.00000       24.00000   
75%           165.00000      165.00000      137.00000       56.00000   
max          1114.00000      660.00000     1091.00000     2461.00000   

            DepDelay       Distance  
count  1936758.00000  1936758.00000  
mean        43.18518      765.68616  
std         53.40250      574.47965  
min          6.00000       11.00000  
25%         12.00000      338.00000  
50%         24.00000      606.00000  
75%         53.00000      998.00000  
max       2467.00000     4962.00000

In [7]:
#  Elimino las filas donde 'AirTime' sea 0, porque luego dará error 

retraso_vuelos.drop(retraso_vuelos[retraso_vuelos['AirTime'] == 0].index, inplace= True)
retraso_vuelos[['ActualElapsedTime', 'CRSElapsedTime','AirTime','ArrDelay', 'DepDelay', 'Distance']].describe().apply(lambda s: s.apply('{0:.5f}'.format)) 

ActualElapsedTime CRSElapsedTime        AirTime       ArrDelay  \
count     1928364.00000  1936553.00000  1928364.00000  1928364.00000   
mean          133.30596      134.30287      108.27754       42.19976   
std            72.06012       71.34147       68.64242       56.78466   
min            14.00000      -25.00000        1.00000     -109.00000   
25%            80.00000       82.00000       58.00000        9.00000   
50%           116.00000      116.00000       90.00000       24.00000   
75%           165.00000      165.00000      137.00000       56.00000   
max          1114.00000      660.00000     1091.00000     2461.00000   

            DepDelay       Distance  
count  1936751.00000  1936751.00000  
mean        43.18508      765.68718  
std         53.40246      574.47998  
min          6.00000       11.00000  
25%         12.00000      338.00000  
50%         24.00000      606.00000  
75%         53.00000      998.00000  
max       2467.00000     4962.00000

### Análisis estadístico

**AirTime**
- 1) Tiempo medio de vuelo : 107,42 minutos 
- 2) Gracias al valor obtenido de desviación estándar se puede concluir que los datos no están muy dispersos
- 3) El 50% de los vuelos tiene una duración de 90 minutos
- 4) Los valores mínimos y máximos problabemente tienen errores porque el vuelo más corto es de 1 minuto y el más largo de 1091 minutos o sea 18,18 h.

**Arrdelay**
- 1) La media de retraso en la hora de llegada son 63,29 minutos

**DepDelay**
- 1) La media de retraso en la hora de salida son 59,67 minutos

**Distance**
- 1) La distancia media que recorren los vuelos es de 559,36 millas 
- 2) La distancia del 50% de los vuelos es de 595 millas

### Datos faltantes por columna

In [8]:
retraso_vuelos.isnull().sum()

Month                     0
DayofMonth                0
DayOfWeek                 0
DepTime                   0
CRSDepTime                0
ArrTime                7110
CRSArrTime                0
UniqueCarrier             0
ActualElapsedTime      8387
CRSElapsedTime          198
AirTime                8387
ArrDelay               8387
DepDelay                  0
Distance                  0
TaxiIn                 7110
TaxiOut                 455
Cancelled                 0
CancellationCode          0
Diverted                  0
CarrierDelay         689268
WeatherDelay         689268
NASDelay             689268
SecurityDelay        689268
LateAircraftDelay    689268
dtype: int64

### Eliminamos filas con datos nulos

In [17]:
retraso_vuelos.dropna(inplace=True)
retraso_vuelos.shape

(1247483, 29)

### Crea columnes noves (velocitat mitjana del vol)

In [10]:
retraso_vuelos['Distance_Km'] = retraso_vuelos['Distance'] * 1.60934 # conversión millas a km
retraso_vuelos['AirTime'].replace(0, np.nan, inplace = True)
retraso_vuelos['AirTime_H'] = retraso_vuelos['AirTime'] / 60 #conversión minutos a h.
retraso_vuelos['Average_Speed'] = retraso_vuelos['Distance_Km'] / retraso_vuelos['AirTime_H'] #columna velocidad media
retraso_vuelos[['Distance', 'AirTime', 'Average_Speed']].head()

Distance  AirTime  Average_Speed
4        515     77.0     645.826052
6       1591    230.0     667.946071
11       828    107.0     747.215058
16      1489    213.0     675.016130
18       838    110.0     735.614684

In [23]:
retraso_vuelos['Average_Speed'].mean()

624.690443650376

### Crea columnas nuevas (si ha llegado tarde)

In [11]:
retraso_vuelos['Late_Landing'] = retraso_vuelos['ArrDelay'] > 0
retraso_vuelos[['UniqueCarrier','Late_Landing']].head(10)

UniqueCarrier  Late_Landing
4             WN          True
6             WN          True
11            WN          True
16            WN          True
18            WN          True
19            WN          True
21            WN          True
26            WN          True
30            WN          True
37            WN          True

### Crea columnas nuevas (si ha salido tarde)

In [12]:
retraso_vuelos['Late_Take_Off'] = retraso_vuelos['DepDelay'] > 0
retraso_vuelos[['Distance', 'AirTime', 'Average_Speed', 'Late_Landing', 'Late_Take_Off']].head(10)

Distance  AirTime  Average_Speed  Late_Landing  Late_Take_Off
4        515     77.0     645.826052          True           True
6       1591    230.0     667.946071          True           True
11       828    107.0     747.215058          True           True
16      1489    213.0     675.016130          True           True
18       838    110.0     735.614684          True           True
19       220     49.0     433.536490          True           True
21       220     47.0     451.984851          True           True
26      1093    143.0     738.045575          True           True
30       972    134.0     700.423200          True           True
37       765    118.0     626.005983          True           True

### Tabla de las aerolínies con más retrasos acumulados.

In [24]:
retrasos_acumulados = retraso_vuelos['Late_Landing'] == True
retraso_vuelos[retrasos_acumulados]['UniqueCarrier'].value_counts()

WN    203559
AA    132257
MQ     97555
UA     95465
OO     88987
DL     72252
XE     72007
US     59508
CO     58958
EV     56781
NW     54744
YV     50646
FL     46991
OH     39293
B6     38232
9E     35631
AS     24012
F9     15940
HA      4325
AQ       340
Name: UniqueCarrier, dtype: int64

### Vuelos más largos

In [14]:
vuelos_largos = retraso_vuelos.sort_values(by= ['Distance'], ascending = False)
vuelos_largos[:10] # muestra los 10 primeros vuelos más largos

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
2363405      4          10          4   1406.0        1335   1837.0   
2345571      4           4          5   1400.0        1335   1830.0   
2355176      4          30          3   1408.0        1335   1843.0   
2354645      4           3          4   1357.0        1335   1839.0   
2353867      4          24          4   1351.0        1335   1900.0   
2353671      4          25          5   1345.0        1335   1826.0   
2353588      4           6          7   1343.0        1335   1831.0   
6980183     12          19          5   1407.0        1335   2217.0   
6980182     12          19          5      5.0        2055   1433.0   
2351025      4          17          4   1508.0        1335   1938.0   

         CRSArrTime UniqueCarrier  ActualElapsedTime  CRSElapsedTime  ...  \
2363405        1810            CO              631.0           635.0  ...   
2345571        1810            CO              630.0           635.0  ...   
2355176        1810            CO              635.0           635.0  ...   
2354645        1810            CO              642.0           635.0  ...   
2353867        1810            CO              669.0           635.0  ...   
2353671        1810            CO              641.0           635.0  ...   
2353588        1810            CO              648.0           635.0  ...   
6980183        1935            CO              790.0           660.0  ...   
6980182        1120            CO              568.0           565.0  ...   
2351025        1810            CO              630.0           635.0  ...   

         CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
2363405          11.0           0.0       0.0            0.0   
2345571           0.0           0.0       0.0            0.0   
2355176          33.0           0.0       0.0            0.0   
2354645           5.0           0.0       7.0            0.0   
2353867           0.0           0.0      34.0            0.0   
2353671           0.0           0.0       6.0            0.0   
2353588           8.0           0.0      13.0            0.0   
6980183           0.0          32.0     130.0            0.0   
6980182          18.0           0.0       3.0            0.0   
2351025          88.0           0.0       0.0            0.0   

         LateAircraftDelay  Distance_Km  AirTime_H Average_Speed  \
2363405               16.0   7985.54508  10.083333    791.954884   
2345571               20.0   7985.54508  10.150000    786.753210   
2355176                0.0   7985.54508  10.250000    779.077569   
2354645               17.0   7985.54508  10.316667    774.043142   
2353867               16.0   7985.54508  10.600000    753.353309   
2353671               10.0   7985.54508  10.333333    772.794685   
2353588                0.0   7985.54508  10.133333    788.047212   
6980183                0.0   7985.54508  10.566667    755.729818   
6980182              172.0   7985.54508   8.866667    900.625385   
2351025                0.0   7985.54508  10.016667    797.225798   

         Late_Landing  Late_Take_Off  
2363405          True           True  
2345571          True           True  
2355176          True           True  
2354645          True           True  
2353867          True           True  
2353671          True           True  
2353588          True           True  
6980183          True           True  
6980182          True           True  
2351025          True           True  

[10 rows x 29 columns]

### Vuelos más retrasados 

In [15]:
vuelos_mas_retrasados = retraso_vuelos.sort_values(by = ['ArrDelay'], ascending = False) #ordeno los vuelos en forma descendente y muestro los 20 vuelos más retrasados
print(vuelos_mas_retrasados[:20]) 

         Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
1018798      2           3          7   1117.0        1820   2256.0   
2235378      4          10          4    724.0        1417    858.0   
2832617      5           6          2      2.0        1530    254.0   
3387883      6          20          5   2140.0        1710   2252.0   
6857047     12          19          5   1602.0        1325   1921.0   
5232546      9           1          1    906.0         714   1048.0   
2232494      4          27          7   1818.0        1633   1942.0   
527950       1           5          6    800.0        1045   1452.0   
4061361      7          16          3    820.0         702   1325.0   
1634129      3          21          5    705.0         615    811.0   

         CRSArrTime UniqueCarrier  ActualElapsedTime  CRSElapsedTime  ...  \
1018798         555            NW              459.0           455.0  ...   
2235378        1605            NW              154.0           1

In [26]:
vuelos_por_dia = retraso_vuelos['DayOfWeek'].value_counts().sort_values(ascending = False)
vuelos_por_dia

5    214235
4    186860
1    186859
7    185183
2    169908
3    167704
6    136734
Name: DayOfWeek, dtype: int64

In [ ]:
#info de interes: como hacer un análisis exploratorio ---> https://www.codificandobits.com/blog/analisis-exploratorio-de-datos/

## Exercici 3

Exporta el dataset net i amb les noves columnes a Excel.

In [32]:
retraso_vuelos.to_csv('ok_retraso_vuelos.csv')

#Corto el DF porque no entra en excel

ok_retraso_vuelos1 = retraso_vuelos.loc[0:1000000].copy()

ok_retraso_vuelos1

ok_retraso_vuelos1.to_excel('retraso_vuelos_corto.xlsx', index=False)